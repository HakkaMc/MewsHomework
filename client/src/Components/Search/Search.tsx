import React from "react";
import { compose } from "recompose";
import { reduxForm } from "redux-form";
import { connect } from "react-redux";

import { actions } from "../../reducers";
import { InputBase, Input } from "../index";

import styles from "./styles.scss";

export interface OuterProps {
	placeholder?: string;
	light?: boolean;
}

export interface Props extends OuterProps {
	searchMovies: typeof actions.movies.searchMovies;
}

export class Search extends React.Component<Props> {
	searchMoviesTimeoutRef: any;

	searchMoviesWrapper = (event: any, newValue: string) => {
		clearTimeout(this.searchMoviesTimeoutRef);
		this.searchMoviesTimeoutRef = setTimeout(() => {
			this.props.searchMovies(newValue);
		}, 1000);
	};

	render() {
		let inputElement;

		if (this.props.light) {
			inputElement = (
				<InputBase
					name="movieName"
					placeholder={this.props.placeholder || "Searchâ€¦"}
					classes={{
						root: styles.inputRoot,
						input: styles.inputInput
					}}
					inputProps={{ "aria-label": "search" }}
					onChange={this.searchMoviesWrapper}
				/>
			);
		} else {
			inputElement = (
				<Input
					name="movieName"
					label={this.props.placeholder}
					onChange={this.searchMoviesWrapper}
				/>
			);
		}

		return (
			<form noValidate autoComplete="off" className={styles.form}>
				{inputElement}
			</form>
		);
	}
}

export default compose<Props, OuterProps>(
	reduxForm({
		form: "search",
		destroyOnUnmount: false
	}),
	connect(null, {
		searchMovies: actions.movies.searchMovies
	})
)(Search);
