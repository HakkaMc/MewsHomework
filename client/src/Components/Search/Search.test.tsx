import React from "react";
import { shallow } from "enzyme";

import { InputBase, Input } from "../index";
import { Search, Props, OuterProps } from "./Search";

describe("<Search/>", () => {
	let props: Props & OuterProps;

	beforeEach(() => {
		props = {
			searchMovies: jest.fn()
		};
	});

	it("Match snapshot width light: false", () => {
		props.light = false;
		const wrapper = shallow(<Search {...props} />);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find(Input)).toHaveLength(1);
		expect(wrapper.find(InputBase)).toHaveLength(0);
	});

	it("Match snapshot width light: true", () => {
		props.light = true;
		const wrapper = shallow(<Search {...props} />);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find(Input)).toHaveLength(0);
		expect(wrapper.find(InputBase)).toHaveLength(1);
	});

	it("Changing input value calls searchMoviesWrapper", () => {
		jest.useFakeTimers();
		props.light = false;
		const wrapper = shallow(<Search {...props} />);
		wrapper.find(Input).prop("onChange")(null, "movie");
		jest.runAllTimers();
		expect(props.searchMovies).toHaveBeenCalledWith("movie");
	});
});
