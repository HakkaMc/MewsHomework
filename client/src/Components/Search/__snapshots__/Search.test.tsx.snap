// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> Match snapshot width light: false 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <Input
    name="movieName"
    onChange={[Function]}
  />
</form>
`;

exports[`<Search/> Match snapshot width light: true 1`] = `
<form
  autoComplete="off"
  noValidate={true}
>
  <Input
    classes={
      Object {
        "input": undefined,
        "root": undefined,
      }
    }
    inputProps={
      Object {
        "aria-label": "search",
      }
    }
    name="movieName"
    onChange={[Function]}
    placeholder="Searchâ€¦"
  />
</form>
`;
